### PR TITLE
PHPLIB-463: Implement spec test runner for retryable reads

### PR DESCRIPTION
**Pull request by alcaeus**
https://jira.mongodb.org/browse/PHPLIB-463

### CODE REVIEW

File: tests/SpecTests/FunctionalTestCase.php [hunk 1]
```diff
@@ -185,31 +185,55 @@ protected function dropTestAndOutcomeCollections()
     {
         $context = $this->getContext();
 
-        $collection = $context->getCollection();
-        $collection->drop($context->defaultWriteOptions);
+        if ($context->bucketName !== null) {
+            $bucket = $context->getGridFSBucket();
```
Comment (jmikola) [inline tests/SpecTests/FunctionalTestCase.php]: `Bucket::drop()` currently does not support any options. This may change in [SPEC-921](http://jira.mongodb.org/browse/SPEC-921) when session support is introduced, but I don't expect it (or any other GridFS methods) will support a write concern option like the Database or Collection drop methods do.

We can consider passing `$context->defaultWriteOptions` to `getGridFSBucket()` here to ensure a majority write concern is used for the ensuing `drop()` operation.

I'll note that only the retryable writes and transactions spec test runners care about using a majority write concern, so we wouldn't need to define a majority write concern in `Context::fromRetryableReads()` as we do in the other spec methods.

Note: This relates to my earlier comment in `getGridFSBucket()`.

I'm OK with whatever you'd like to do here.

File: src/Operation/MapReduce.php [hunk 2]
```diff
@@ -264,7 +264,7 @@ public function execute(Server $server)
 
         $cursor = $hasOutputCollection
             ? $server->executeReadWriteCommand($this->databaseName, $command, $options)
-            : $server->executeReadCommand($this->databaseName, $command, $options);
+            : $server->executeCommand($this->databaseName, $command, $options);
```
[the remark this answers is not in the document]
Comment (alcaeus) [reply src/Operation/MapReduce.php]: Please correct me if I'm wrong, but [Server::executeCommand](https://github.com/mongodb/mongo-php-driver/blob/e65d114b405ea0fd3f85eab6ef113d782d90c1fc/src/MongoDB/Server.c#L50) uses `PHONGO_COMMAND_RAW` (`0x07`). The way I read the `readConcern` check in [`phongo_execute_command`](https://github.com/mongodb/mongo-php-driver/blob/e65d114b405ea0fd3f85eab6ef113d782d90c1fc/php_phongo.c#L949), it will parse the read concern when using `PHONGO_COMMAND_RAW`.

However, when looking at `mongoc-client-c`, [it only applies a `readConcern` option for `MONGOC_CMD_READ` and `MONGOC_CMD_RW`](https://github.com/mongodb/mongo-c-driver/blob/b37f263e2286ad20c656df1e96fa106377fd7c7a/src/libmongoc/src/mongoc/mongoc-client.c#L1982).

Looking at the `enum` we declare in [`phongo.h`](https://github.com/mongodb/mongo-php-driver/blob/e65d114b405ea0fd3f85eab6ef113d782d90c1fc/php_phongo.h#L120), I believe the declaration for `PHONGO_COMMAND_RAW` should change from `0x07` to `0x06` to not support the `readConcern` option, just the way libmongoc does.

In summary, your point is absolutely correct, as libmongoc does not parse a `readConcern` option when running the command through `executeCommand` vs. `executeReadCommand`. Given the history of `mapReduce`, I'd also be surprised if it started supporting a `readConcern` option. If that were the case however, there would be more changes necessary in both libmongoc and the driver here, so I think we should be safe ignoring this for now. Regardless of the above, I've added a comment to explain why we use `executeCommand` instead of `executeReadCommand`.

File: tests/SpecTests/Context.php [hunk 3]
```diff
@@ -143,6 +169,11 @@ public function getDatabase(array $databaseOptions = [])
         return $this->selectDatabase($this->databaseName, $databaseOptions);
     }
 
+    public function getGridFSBucket(array $bucketOptions = [])
```
[the remark this answers is not in the document]
Comment (alcaeus) [reply tests/SpecTests/Context.php]: I mainly added the options for consistency with `getCollection`. I've added `$context->defaultWriteOptions` when constructing the bucket; this was an oversight. Thanks for spotting it 👍

File: tests/SpecTests/FunctionalTestCase.php [hunk 4]
```diff
@@ -220,6 +244,17 @@ private function getOutcomeCollection()
         return new Collection($this->manager, $context->databaseName, $context->outcomeCollectionName);
     }
 
+    private function insertData($documents, $collection, array $writeOptions)
```
[the remark this answers is not in the document]
Comment (alcaeus) [reply tests/SpecTests/FunctionalTestCase.php]: That is correct. I've applied your suggestion from below 👍

File: tests/SpecTests/Operation.php [hunk 5]
```diff
@@ -214,6 +239,40 @@ private function execute(Context $context)
...
+        switch ($this->name) {
+            case 'listDatabaseNames':
+                return array_map(
+                    function (DatabaseInfo $info) {
+                        return $info->getName();
+                    },
+                    iterator_to_array($client->listDatabases())
+                );
+
+            case 'listDatabaseObjects':
```
[the remark this answers is not in the document]
Comment (alcaeus) [reply tests/SpecTests/Operation.php]: We have a few options to skip these tests:
* dropping the respective files from the spec tests folder
* skipping them via the `skipReason` field in the spec file itself.
* skipping them in the spec test itself if the test name begins with one of the skipped operation names

Of all these, the last option is the most resilient when we sync spec tests in the future, as it allows us to copy test files over from the specifications directory without having to worry about changing them or skipping some of them. I agree that we may as well skip them since the entire logic for these operations is contained in the tests.

> So good on you for going the extra mile :)

Credit should go do the C driver team, which [also implemented this the same way](https://github.com/mongodb/mongo-c-driver/blob/b37f263e2286ad20c656df1e96fa106377fd7c7a/src/libmongoc/tests/json-test-operations.c#L1909). I wasn't sure how to implement the operation, so I looked there for guidance. This also highlights that their implementation has the same flaw as mine, so we may want to also skip these tests there or implement the operation correctly.

File: tests/SpecTests/ResultExpectation.php [hunk 6]
```diff
@@ -85,6 +85,19 @@ public static function fromCrud(stdClass $operation, $defaultAssertionType)
         return new self($assertionType, $expectedValue);
     }
 
+    public static function fromRetryableReads(stdClass $operation, $defaultAssertionType)
+    {
+        if (property_exists($operation, 'result') && !self::isErrorResult($operation->result)) {
```
[the remark this answers is not in the document]
Comment (alcaeus) [reply tests/SpecTests/ResultExpectation.php]: Good point. I've change the default assertion type to `ASSERT_SAME` and return the contents of the GridFS file in the operation. 👍

File: tests/SpecTests/Context.php [hunk 7]
```diff
@@ -272,6 +299,20 @@ public function selectDatabase($databaseName, array $databaseOptions = [])
...
 
+    public function selectGridFSBucket($databaseName, $bucketName, array $bucketOptions = [])
+    {
+        return $this->selectDatabase($databaseName)->selectGridFSBucket($this->prepareGridFSBucketOptions($bucketOptions, $bucketName));
+    }
+
+    private function prepareGridFSBucketOptions(array $options, $bucketPrefix)
+    {
+        if ($bucketPrefix !== null) {
+            $options['bucketPrefix'] = $bucketPrefix;
```
[the remark this answers is not in the document]
Comment (alcaeus) [reply tests/SpecTests/Context.php]: The GridFS tests contain a `bucket_name` field, which is read in the spec test and passed to the context. This is used when calling `getGridFSBucket`, which is the counterpart to `getDatabase` and `getCollection`.

File: tests/SpecTests/Operation.php [hunk 8]
```diff
@@ -443,9 +581,12 @@ private function getResultAssertionTypeForDatabase()
     {
         switch ($this->name) {
             case 'aggregate':
+            case 'listCollections':
                 return ResultExpectation::ASSERT_SAME_DOCUMENTS;
             case 'runCommand':
                 return ResultExpectation::ASSERT_MATCHES_DOCUMENT;
+            case 'watch':
+                return ResultExpectation::ASSERT_NOTHING;
```
[the remark this answers is not in the document]
Comment (alcaeus) [reply tests/SpecTests/Operation.php]: I missed that the original comment referred to `match` as well. It's updated now.

File: tests/SpecTests/Operation.php [hunk 9]
```diff
@@ -337,16 +404,61 @@ private function executeForDatabase(Database $database, Context $context)
...
+     * Executes the GridFS bucket operation and return its result.
+     *
+     * @param Bucket  $bucket
+     * @param Context $context Execution context
+     * @return mixed
+     * @throws LogicException if the database operation is unsupported
+     */
+    private function executeForGridFSBucket(Bucket $bucket, Context $context)
+    {
+        switch ($this->name) {
```
[the remark this answers is not in the document]
Comment (alcaeus) [reply tests/SpecTests/Operation.php]: > I suppose there's no reason to do this because we reference exact options in the cases below and aren't passing the full array to anything. I suppose it's something to keep in mind if/when GridFS gets session support, although who knows what that API may look like.

That is correct. We only use the `filename` option for the `download_by_name` operation and leave all other options unused. We could pass in options on the off-chance that a test specifies the `revision` option (which currently is the only option taken by these GridFS operations).

**Comment by alcaeus** [comment]
Merged in https://github.com/mongodb/mongo-php-library/commit/27892777027c7015da3be75f69978dabeabedcf7.
